server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+
+        MockPublication(Settings settings, PublishRequest publishRequest, Discovery.AckListener ackListener,
+                               LongSupplier currentTimeSupplier) {
+            super(settings, publishRequest, ackListener, currentTimeSupplier);
+            this.publishRequest = publishRequest;
+        }
+
+        @Override
+        protected void onCompletion(boolean committed) {
+            completed = true;
+            this.committed = committed;
+        }
+
+        @Override
+        protected void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response) {
+
+        }
+
+        @Override
+        protected void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
+                                          ActionListener<PublishWithJoinResponse> responseActionListener) {
+            assertSame(publishRequest, this.publishRequest);
+            assertNull(pendingPublications.put(destination, responseActionListener));
+        }
+
+        @Override
+        protected void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
+                                       ActionListener<TransportResponse.Empty> responseActionListener) {
+            if (this.applyCommit == null) {
+                this.applyCommit = applyCommit;
+            } else {
+                assertSame(applyCommit, this.applyCommit);
+            }
+            assertNull(pendingCommits.put(destination, responseActionListener));
+        }
+    }
+
+    DiscoveryNode n1 = CoordinationStateTests.createNode("node1");
+    DiscoveryNode n2 = CoordinationStateTests.createNode("node2");
+    DiscoveryNode n3 = CoordinationStateTests.createNode("node3");
+    Set<DiscoveryNode> discoNodes = Sets.newHashSet(n1, n2, n3);
+
+    MockNode node1 = new MockNode(Settings.EMPTY, n1);
+    MockNode node2 = new MockNode(Settings.EMPTY, n2);
+    MockNode node3 = new MockNode(Settings.EMPTY, n3);
+    List<MockNode> nodes = Arrays.asList(node1, node2, node3);
+
+    Function<DiscoveryNode, MockNode> nodeResolver = dn -> nodes.stream().filter(mn -> mn.localNode.equals(dn)).findFirst().get();
+
+    private void initializeCluster(VotingConfiguration initialConfig) {
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
```

Comment:
Could this make finer assertions about the state that the publication is in? More precise requests follow...

Reply:
see comments below